src/query/functions/binary/node.go

```diff
@@ -0,0 +1,482 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package binary
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/functions/logical"
+	"github.com/m3db/m3/src/query/parser"
+)
+
+const (
+	// *******************************************
+	// * Arithmetic functions
+
+	// PlusType adds datapoints in both series
+	PlusType = "+"
+
+	// MinusType subtracts rhs from lhs
+	MinusType = "-"
+
+	// MultiplyType multiplies datapoints by series
+	MultiplyType = "*"
+
+	// DivType divides datapoints by series
+	// Special cases are:
+	// 	 X / 0 = +Inf
+	// 	-X / 0 = -Inf
+	// 	 0 / 0 =  NaN
+	DivType = "/"
+
+	// ExpType raises lhs to the power of rhs
+	// NB: to keep consistency with prometheus (and go)
+	//  0 ^ 0 = 1
+	//  NaN ^ 0 = 1
+	ExpType = "^"
+
+	// ModType takes the modulo of lhs by rhs
+	// Special cases are:
+	// 	 X % 0 = NaN
+	// 	 NaN % X = NaN
+	// 	 X % NaN = NaN
+	ModType = "%"
+
+	// *******************************************
+	// * Comparison functions
+
+	// EqType raises lhs to the power of rhs
+	EqType = "=="
+
+	// NotEqType raises lhs to the power of rhs
+	NotEqType = "!="
+
+	// GreaterType raises lhs to the power of rhs
+	GreaterType = ">"
+
+	// LesserType raises lhs to the power of rhs
+	LesserType = "<"
+
+	// GreaterEqType raises lhs to the power of rhs
+	GreaterEqType = ">="
+
+	// LesserEqType raises lhs to the power of rhs
+	LesserEqType = "<="
+)
+
+type mathFunc func(x, y float64) float64
+
+type singleScalarFunc func(x float64) float64
+
+func toFloat(b bool) float64 {
+	if b {
+		return 1
+	}
+	return 0
+}
+
+var (
+	mathFuncs = map[string]mathFunc{
+		PlusType:     func(x, y float64) float64 { return x + y },
+		MinusType:    func(x, y float64) float64 { return x - y },
+		MultiplyType: func(x, y float64) float64 { return x * y },
+		DivType:      func(x, y float64) float64 { return x / y },
+		ModType:      math.Mod,
+		ExpType:      math.Pow,
+
+		EqType:        func(x, y float64) float64 { return toFloat(x == y) },
+		NotEqType:     func(x, y float64) float64 { return toFloat(x != y) },
+		GreaterType:   func(x, y float64) float64 { return toFloat(x > y) },
+		LesserType:    func(x, y float64) float64 { return toFloat(x < y) },
+		GreaterEqType: func(x, y float64) float64 { return toFloat(x >= y) },
+		LesserEqType:  func(x, y float64) float64 { return toFloat(x <= y) },
+	}
+
+	errLeftScalar              = errors.New("expected left scalar but node type incorrect")
+	errRightScalar             = errors.New("expected right scalar but node type incorrect")
+	errNoModifierForComparison = errors.New("comparisons between scalars must use BOOL modifier")
+	errNoMatching              = errors.New("functions must have a vector matching set")
```

Comment:
can you make the message clear ? Not sure what a vector matching set is

Reply:
Sure, it does seem a bit weird. It's more of a sanity check since I expect a parse error before it gets here, but will make it clearer